internal/broker/bind.go

```diff
@@ -3,36 +3,328 @@ package broker
 	"fmt"
+	"net/http"
+	"sync"
 
+	jsonhash "github.com/komkom/go-jsonhash"
 	"github.com/pkg/errors"
 	osb "github.com/pmorie/go-open-service-broker-client/v2"
+	"github.com/sirupsen/logrus"
+	"k8s.io/helm/pkg/proto/hapi/chart"
 
 	"github.com/kyma-project/helm-broker/internal"
 )
 
 type bindService struct {
-	instanceBindDataGetter instanceBindDataGetter
+	addonIDGetter                 addonIDGetter
+	chartGetter                   chartGetter
+	instanceGetter                instanceGetter
+	bindTemplateRenderer          bindTemplateRenderer
+	bindTemplateResolver          bindTemplateResolver
+	instanceBindDataInserter      instanceBindDataInserter
+	instanceBindDataGetter        instanceBindDataGetter
+	instanceBindDataRemover       instanceBindDataRemover
+	bindStateGetter               bindStateBindingGetter
+	bindOperationGetter           bindOperationGetter
+	bindOperationCollectionGetter bindOperationCollectionGetter
+	bindOperationInserter         bindOperationInserter
+	bindOperationUpdater          bindOperationUpdater
+	operationIDProvider           func() (internal.OperationID, error)
+	mu                            sync.Mutex
+
+	log *logrus.Entry
+
+	testHookAsyncCalled func(internal.OperationID)
 }
 
-func (svc *bindService) Bind(ctx context.Context, osbCtx OsbContext, req *osb.BindRequest) (*osb.BindResponse, error) {
+func (svc *bindService) Bind(ctx context.Context, osbCtx OsbContext, req *osb.BindRequest) (*osb.BindResponse, *osb.HTTPStatusCodeError) {
 	if len(req.Parameters) > 0 {
-		return nil, fmt.Errorf("helm-broker does not support configuration options for the service binding")
+		return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusBadRequest, ErrorMessage: strPtr("helm-broker does not support configuration options for the service binding")}
+	}
+
+	if !req.AcceptsIncomplete {
+		return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusBadRequest, ErrorMessage: strPtr("asynchronous operation mode required")}
+	}
+
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+
+	iID := internal.InstanceID(req.InstanceID)
+	bID := internal.BindingID(req.BindingID)
+	svcID := internal.ServiceID(req.ServiceID)
+	svcPlanID := internal.ServicePlanID(req.PlanID)
+	paramHash := jsonhash.HashS(req.Parameters)
+
+	switch opIDInProgress, inProgress, err := svc.bindStateGetter.IsBindingInProgress(iID, bID); true {
+	case err != nil:
+		return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusBadRequest, ErrorMessage: strPtr(fmt.Sprintf("while checking if service binding is being created: %v", err))}
+	case inProgress:
+		opKeyInProgress := osb.OperationKey(opIDInProgress)
+		return &osb.BindResponse{Async: true, OperationKey: &opKeyInProgress}, nil
+	}
+
+	switch bindOp, state, err := svc.bindStateGetter.IsBound(iID, bID); true {
+	case err != nil:
+		return nil, &osb.HTTPStatusCodeError{StatusCode: http.StatusBadRequest, ErrorMessage: strPtr(fmt.Sprintf("while checking if service binding for service instance already exists: %v", err))}
```

Comment:
is the StatusBadRequest ok?

Reply:
This should be StatusInternalServerError, because the only way IsBound method returns error is problem with calling database (e.g. etcd not working)